File: Hash Table/java/leetcode347_Top_K_Frequent_Elements.java
//Given a non-empty array of integers, return the k most frequent elements.
//
//Example 1:
//
//Input: nums = [1,1,1,2,2,3], k = 2
//Output: [1,2]
//Example 2:
//
//Input: nums = [1], k = 1
//Output: [1]

package wenxinjie3;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class leetcode347_Top_K_Frequent_Elements {
    public List<Integer> topKFrequent(int[] nums, int k) {
        List<Integer>[] bucket = new List[nums.length + 1];
        Map<Integer, Integer> map = new HashMap<Integer, Integer>();
        
        for (int n : nums){
            map.put(n, map.getOrDefault(n, 0) + 1);
        }
        
        for (int key: map.keySet()){
            int frequency = map.get(key);
            if (bucket[frequency] == null){
                bucket[frequency] = new ArrayList<>();
            }
            bucket[frequency].add(key);
        }
        
        List<Integer> res = new ArrayList<Integer>();
        
        for (int pos = bucket.length - 1; pos >= 0 && res.size() < k; pos--) {
		if (bucket[pos] != null) {
			res.addAll(bucket[pos]);
		}
	}
	return res;
    }
}

//Time: O(n)
//Space: O(n)
//Difficulty: medium       

File: Hash Table/python/leetcode347_Top_K_Frequent_Elements.py
# Given a non-empty array of integers, return the k most frequent elements.

# Example 1:

# Input: nums = [1,1,1,2,2,3], k = 2
# Output: [1,2]
# Example 2:

# Input: nums = [1], k = 1
# Output: [1]

class Solution(object):
    def topKFrequent(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        map1 = collections.defaultdict(int)
        for n in nums:
        	map1[n] += 1

        map2 = collections.defaultdict(list)
        for i,j in map1.items():
        	map2[j].append(i)

        res = []
        keys = sorted(list(map2.keys()))[::-1]
        for i in keys:
        	res.extend(map2[i])
        	if len(res) == k:
        		break
        return res

# Time: O(n)
# Space: O(n)
# Difficulty: medium       
